### PR TITLE
[Add] daily challenges feature

### PR DESCRIPTION
**Pull request by siva2204**
created Config.go in models package to handle marketday and dailychallenge status

### CODE REVIEW

File: migrations/20210127061413_create_config_table.down.sql [hunk 1]
```diff
@@ -0,0 +1 @@
+DROP TABLE IF EXISTS Config;
```
Comment (PranavShriram) [inline migrations/20210127061413_create_config_table.down.sql:1]: Add newline at end of file

File: migrations/20210127061413_create_config_table.up.sql [hunk 2]
```diff
@@ -0,0 +1,7 @@
+CREATE TABLE IF NOT EXISTS Config (
+    id int(11) UNSIGNED NOT NULL, 
+    isDailyChallengeOpen BOOLEAN NOT NULL DEFAULT FALSE,
+    marketDay int(11) UNSIGNED NOT NULL DEFAULT 0,
+    isMarketOpen BOOLEAN NOT NULL DEFAULT FALSE,
+    PRIMARY KEY (id)
+);
```
Comment (PranavShriram) [inline migrations/20210127061413_create_config_table.up.sql:7]: Add newline at end of file

File: migrations/20210117174003_create_user_state_daily_challenge_table.sql.up.sql [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+     `userId` int(11) UNSIGNED NOT NULL,
+     `marketDay` int(11) UNSIGNED NOT NULL,
+     `initialValue`bigint(11) SIGNED NOT NULL,
+     `finalValue`bigint(11) SIGNED DEFAULT NULL,
+     `isCompleted`BOOLEAN,
+     `isRewardClaimed`BOOLEAN,
+     PRIMARY KEY (id),
+     FOREIGN KEY (challengeId) REFERENCES DailyChallenge(id),
+     FOREIGN KEY (userId) REFERENCES Users(id)
+) AUTO_INCREMENT=1;
```
Comment (PranavShriram) [inline migrations/20210117174003_create_user_state_daily_challenge_table.sql.up.sql:13]: Add newline at end of file

File: migrations/20210117174003_create_user_state_daily_challenge_table.sql.down.sql [hunk 4]
```diff
@@ -0,0 +1 @@
+DROP TABLE IF EXISTS UserState;
```
Comment (PranavShriram) [inline migrations/20210117174003_create_user_state_daily_challenge_table.sql.down.sql:1]: Add newline at end of file

File: migrations/20210116114500_create_daily_challenge_table.sql.up.sql [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+    `id` int(11) UNSIGNED NOT NULL AUTO_INCREMENT,
+    `marketDay` int(11) UNSIGNED NOT NULL,
+    `challengeType` enum('Cash','NetWorth','StockWorth','SpecificStock'),
+    `value` bigint(11) SIGNED NOT NULL,
+    `stockId` int(11) UNSIGNED DEFAULT NULL,
+    `reward` int(11) UNSIGNED NOT NULL,
+    PRIMARY KEY (id),
+    FOREIGN KEY (stockId) REFERENCES Stocks(id),
+    CONSTRAINT check_daily_challenge CHECK((challengeType='SpecificStock' AND stockId>0 ) OR (challengeType<>'SpecificStock' AND stockId=NULL))
+) AUTO_INCREMENT=1;
```
Comment (PranavShriram) [inline migrations/20210116114500_create_daily_challenge_table.sql.up.sql:11]: Add newline at end of file

File: models/User.go [hunk 6]
```diff
@@ -2444,3 +2444,38 @@ func UnBlockAllUsers() error {
 
 	return nil
 }
+
+//GetUserStockWorth returns total stockworth of a user including reserved StockWorth
+func GetUserStockWorth(userId uint32) (int64, error) {
```
Comment (abesheknarayan) [inline models/User.go:2449]: for getting user net stock worth , we can use Net worth - (cash+reserved cash)

File: models/DailyChallenge.go [hunk 7]
```diff
@@ -0,0 +1,846 @@
...
+		return err
+	}
+
+	//query to get cash,Stockworth,NetWorth for all the non-blocked users
+	query := fmt.Sprintf(`
+	SELECT U.id as user_id, U.cash + U.reservedCash as cash,
+	 ifNull((SUM(cast(S.currentPrice AS signed) * cast(T.stockQuantity AS signed)) + SUM(cast(S.currentPrice AS signed) * cast(T.reservedStockQuantity AS signed)) ),0) AS stock_worth,
+	 ifnull((U.cash + U.reservedCash + SUM(cast(S.currentPrice AS signed) * cast(T.stockQuantity AS signed)) + SUM(cast(S.currentPrice AS signed) * cast(T.reservedStockQuantity AS signed))),U.cash) AS total
+	 FROM
+	Users U LEFT JOIN Transactions T ON U.id = T.userId LEFT JOIN Stocks S ON T.stockId = S.id WHERE U.blockCount < %d GROUP BY U.id;`, config.MaxBlockCount)
```
Comment (abesheknarayan) [inline models/DailyChallenge.go:478]: Can use inner join instead of left join

File: models/DailyChallenge.go [hunk 8]
```diff
@@ -0,0 +1,846 @@
...
+	if err := tx.Error; err != nil {
+		l.Error(err)
+		return err
+	}
+
+	var queryResults []updateUserStateQueryData
+
+	query := fmt.Sprintf(`SELECT U.id AS id,U.userId AS user_id, U.challengeId AS challenge_id,U.initialValue AS initial_value,D.challengeType AS challenge_type,D.value AS value,D.stockId AS stock_id
+	 FROM
+	UserState U LEFT JOIN DailyChallenge D ON U.challengeId = D.id WHERE U.marketDay = %d;`, marketday)
```
Comment (abesheknarayan) [inline models/DailyChallenge.go:300]: Can use inner join instead of left join

**Review by abesheknarayan — APPROVED** [review]

**Pull request merged**